### DISCUSSION TITLE
Go modules private repos: support well-known hosts: gitlab.com, bitbucket.org

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#21671] [Suggest an Idea]
Add support for gitlab.com and bitbucket.org just like has been added for github.com in #3495. i.e. if a token is found for either of the hosts, add it to the docker run command.

Also see #3202 for further discussions

**Comment by dfreilich** [comment]
Is there any movement on this issue? We're self-hosting on Gitlab, and it's failing to do the `postUpdateOptions` of `gomodTidy` because of our private go modules.

**Comment by rarkins** [comment]
Yes, check current PRs

**Comment by pspurlock07** [comment]
Am I correct in understanding this is the PR for adding support for this? https://github.com/renovatebot/renovate/pull/13477

I'm just checking because I saw merged PRs like this https://github.com/renovatebot/renovate/pull/13352 so just want to make sure I'm not doing something wrong in my renovate config, or this feature actually is not implemented yet.

**Comment by Chumper** [comment]
@pspurlock07 yes, that should solve this use case.
#13352 seems more like a github registry problem than a authentication problem that we are trying to solve in #13477

**Comment by pspurlock07** [comment]
I wanted to post in here for more visibility. A side effect of this issue I just realized is that anytime you are using private go mods not supported by renovate, then any MR/update renovate creates will fail to update the go.sum, even for otherwise valid modules. Any other module that isn't private that renovate can update will have an MR generated with this error:

```
⚠ Artifact update problem
...
The artifact failure details are included below:
File name: go.sum

Command failed: go get -d -t ./...
go: downloading github.com/google/go-cmp v0.5.8
go: module gitlab.com/myprivategroup/myprivateproject: git ls-remote -q origin in /go/pkg/mod/cache/vcs/845610008a41f270b2f7971ef244b1950f14675cf0a5a918366fd3eb70535ab8: exit status 128:
	remote: The project you were looking for could not be found or you don't have permission to view it.
	fatal: repository 'https://gitlab.com/myprivategroup.git' not found
```

Renovate will update the `go.mod` file just fine, but is unable to update `go.sum`, so the update can't be used since both need to be updated in tandem.

**Comment by pspurlock07** [comment]
> 

This isn't about private github.com modules. It's about private gitlab modules. The error above shows a gitlab.com module.

My point here is if you try to use a currently unsupported private go module, then it doesn't matter what renovate tries to update at that point, it all fails. In my error above, you can see I have another module that is from github, but it cannot update the go.sum because I have a private gitlab module it fails on.

**Comment by rarkins** [comment]
@pspurlock07 you wrote "anytime you are using private go mods, then renovate essentially can't be used" and this is incorrect, either intentionally or unintentionally. If you don't remove or edit it the exaggerated comment, then it's clearly not unintentional. Please revise your comment to be accurate.

I also disagree that it's pointless to use it, because there are ways you could configure another script to run after Renovate's with appropriate credentials to fix/add the go.sum. But that one is more a matter of opinion.

**Comment by pspurlock07** [comment]
Ok, I edited the comment to say "anytime you are using private go mods not supported by renovate". I was speaking in the context of this issue and was not trying to exaggerate.

I hadn't thought about trying to use scripts to fix the go.sum, but I know from a user standpoint all of the MRs renovate creates for mods that it can update will come with that error I posted above, making user believe they are doing something wrong when they aren't.
